src/Bot.hs

```diff
@@ -24,15 +24,29 @@ app' = do
   S.get "/" $ do
     S.status $ Status 200 "OK"
   S.post "/type" $ do
-    res <- parseFormEncodedBody . toStrict . decodeUtf8 <$> S.body
-    case (removeCommandChars <$> lookup "text" res) of
-      Just x -> do
-        hoogleRes <- liftIO . simpleHttp . functionNameToUrl $ x
-        maybe badRequest (liftIO . slackRequest) (firstType =<< decode hoogleRes)
-      Nothing -> badRequest
+    params <- parseFormEncodedBody . toStrict . decodeUtf8 <$> S.body
+    authorized <- liftIO . withAuthorization $ lookup "token" params
+    case authorized of
+      True -> typeRequest params
+      False -> unauthorized
```

Comment:
case/True/False seems odd instead of an if/then/else. That said, I think there are some broader things I'd consider changing:

- `getEnv` should happen elsewhere, when the app is loaded, not every request. Rather a compile-time error than a run-time error, rather a startup error than an on-request error (I'm assuming whatever `S.ActionM` supports something `Reader`-like to do this.)

- This approach to parameter handling and authorization seem too low level to scale, what do you think about something like this:

  ```hs
  S.post "/type" $ authorized $ do
      text <- requireParameter "text"
      -- I'd probably refactor the hoogle/slack interfaces too, to make this bit
      -- read better, but that can be done separately

  authorized :: S.ActionM () -> S.ActionM ()
  authorized f = do
      settings <- getAppSettings
      mtoken <- lookupParameter "token"

      case mtoken of
          Just t | t == slackToken appSettings -> f
          _ -> unauthorized

  lookupParameter :: Text -> S.Action (Maybe Text)
  lookupParameter = undefined

  requireParameter :: Text -> S.ActionM Text
  requireParameter k = maybe badRequest' return =<< lookupParameter k
    where
      badRequest' = badRequest $ "Parameter " <> k <> " is required"
  ```